Mathlib/GroupTheory/DivisibleHull.lean

```diff
@@ -0,0 +1,301 @@
+    rw [mul_right_comm sc sa sc, mul_right_comm sc sb sc]
+    rw [mul_smul _ _ ma, mul_smul _ _ mb]
+    exact zsmul_le_zsmul_right (le_of_lt (sc * sc).prop) h
+
+instance : PosSMulStrictMono ℚ (DivisibleHull M) where
+  elim a ha b c h := by
+    induction b using ind with | mk mb sb
+    induction c using ind with | mk mc sc
+    simp_rw [mk_lt_mk, Submonoid.smul_def] at h
+    simp_rw [qsmul_mk, mk_lt_mk, Submonoid.smul_def, Submonoid.mul_def, smul_smul]
+    simp_rw [mul_right_comm _ _ a.num, mul_smul _ _ mc, mul_smul _ _ mb]
+    refine (zsmul_lt_zsmul_iff_right ?_).mpr h
+    exact mul_pos (Int.natCast_pos.mpr a.den_pos) (Rat.num_pos.mpr ha)
+
+variable (M) in
+/-- The map `m ↦ m / 1` as an `OrderAddMonoidHom`. -/
+@[simps]
+def mkOrderAddMonoidHom : M →+o DivisibleHull M where
+  toFun a := mk a 1
+  map_add' a b := by simp [mk_add_mk]
+  map_zero' := by simp [mk]
+  monotone' a b h := by simpa [mk_le_mk] using h
+
+variable (M) in
+theorem mkOrderAddMonoidHom_injective : Function.Injective (mkOrderAddMonoidHom M) := by
+  intro a b
+  simp [mk_eq_mk_iff_smul_eq_smul]
+
+variable (M) in
+/-- The `OrderHom` on `ArchimedeanClass` induced by `DivisibleHull.mkOrderAddMonoidHom`. -/
+noncomputable
+def archimedeanClassOrderHom : ArchimedeanClass M →o ArchimedeanClass (DivisibleHull M) :=
+  ArchimedeanClass.orderHom (mkOrderAddMonoidHom M)
+
+theorem archimedeanClassMk_mk (m : M) (s : Submonoid.pos ℤ) :
+    ArchimedeanClass.mk (mk m s) = archimedeanClassOrderHom M (ArchimedeanClass.mk m) := by
+  trans ArchimedeanClass.mk (s.val • mk m s)
+  · rw [ArchimedeanClass.mk_smul _ (ne_of_gt s.prop)]
+  rw [archimedeanClassOrderHom, ArchimedeanClass.orderHom_mk, mkOrderAddMonoidHom_apply, zsmul_mk]
+  apply congrArg
+  simp [mk_eq_mk_iff_smul_eq_smul, Submonoid.smul_def]
+
+variable (M) in
+theorem archimedeanClassOrderHom_injective : Function.Injective (archimedeanClassOrderHom M) :=
+  ArchimedeanClass.orderHom_injective (mkOrderAddMonoidHom_injective M)
+
+variable (M) in
+/-- The inverse of `DivisibleHull.archimedeanClassOrderHom` by taking the `ArchimedeanClass`
+of the numerator. -/
+noncomputable
+def archimedeanClassOrderHomInv : ArchimedeanClass (DivisibleHull M) →o ArchimedeanClass M :=
+  ArchimedeanClass.liftOrderHom (fun x ↦ x.liftOn (fun m s ↦ ArchimedeanClass.mk m)
+    (fun _ _ _ _ h ↦ by
+      apply archimedeanClassOrderHom_injective M
+      apply_fun ArchimedeanClass.mk at h
+      simpa [archimedeanClassMk_mk] using h))
+    (fun a b h ↦ by
+      induction a using ind with | mk _ _
+      induction b using ind with | mk _ _
+      simp_rw [archimedeanClassMk_mk] at h
+      simpa using ((archimedeanClassOrderHom M).monotone.strictMono_of_injective
+        (archimedeanClassOrderHom_injective M)).le_iff_le.mp h)
+
+variable (M) in
+/-- The Archimedean classes of `DivisibleHull M` are the same as those of `M`. -/
+noncomputable
+def archimedeanClassOrderIso : ArchimedeanClass M ≃o ArchimedeanClass (DivisibleHull M) :=
```

Comment:
Assuming that whole initiative about making Mathlib build by replacing `by` with `sorry` is no longer extant, I think the code would look nicer if you used `apply` for the first line.